### PR TITLE
Feat/add numerical utils for divide mod

### PR DESCRIPTION
**Pull request by stuarttimwhite**
Please be sure to look over the pull request guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

# Please go through the following checklist
- [ ] The PR title and commit messages adhere to guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md. In particular `!` is used if and only if at least one breaking change has been introduced.
- [ ] I have run the ci check script with `source scripts/run_ci_checks.sh`.

# Rationale for this change

In preparation for the new DivideExpr and ModuloExpr, some utilities are needed.

# What changes are included in this PR?

New divide and modulo utility functions, as well as a couple other convenience functions.

# Are these changes tested?
Yes. Some tests for edge cases and operations involving differing types.

### CODE REVIEW

**Review by JayWhite2357 — COMMENTED** [review]
A few NITs.
In general, this might belong in `column_arithmetic_operation` or similar.
Also, take a look at how `column_arithmetic_operation` reuses code. It could reduce boilerplate here.

**Comment by stuarttimwhite** [comment]
> A few NITs. In general, this might belong in `column_arithmetic_operation` or similar. Also, take a look at how `column_arithmetic_operation` reuses code. It could reduce boilerplate here.

Do you mean the entirety of `numerical_util` might belong in `column_arithmetic_operation`? It seems to me that `divide_columns` and `modulo_columns` have similar signatures and purposes to `multiply_expr` for example, so I think they should be in the same place. Admittedly, the divide and modulo code is far bigger, thanks to the matching statements.

**Comment by JayWhite2357** [comment]
> > A few NITs. In general, this might belong in `column_arithmetic_operation` or similar. Also, take a look at how `column_arithmetic_operation` reuses code. It could reduce boilerplate here.
> 
> Do you mean the entirety of `numerical_util` might belong in `column_arithmetic_operation`? It seems to me that `divide_columns` and `modulo_columns` have similar signatures and purposes to `multiply_expr` for example, so I think they should be in the same place. Admittedly, the divide and modulo code is far bigger, thanks to the matching statements.

Possibly, also just the trick of how it uses traits so that the match only needs to be written once.

**Review by JayWhite2357 — COMMENTED** [review]
Never hit submit on this review, sorry.

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 1]
```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
...
+            assert_eq!(quotient, (numerator, scalar_column));
+        }
+    }
+
+    #[test]
+    fn we_can_modulo_columns_for_each_type() {
+        let alloc = Bump::new();
+        let tiny_int_column: Column<
+            '_,
+            crate::base::scalar::MontScalar<crate::base::scalar::test_scalar::TestMontConfig>,
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:662]:
```suggestion
            TestScalar,
```

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 2]
```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
...
+        let expected_scalars = [TestScalar::from(3), TestScalar::from(-6)];
+        assert_eq!(columns_to_scalar_slice(&column, &alloc), expected_scalars);
+    }
+
+    #[test]
+    fn we_can_divide_columns_for_each_type() {
+        let alloc = Bump::new();
+        let tiny_int_column: Column<
+            '_,
+            crate::base::scalar::MontScalar<crate::base::scalar::test_scalar::TestMontConfig>,
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:637]:
```suggestion
            TestScalar,
```

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 3]
```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
...
+            tiny_int_column,
+            small_int_column,
+            int_column,
+            big_int_column,
+            int128_column,
+        ];
+        let columns = columns.iter().enumerate();
+        for ((ni, numerator), (di, denominator)) in iproduct!(columns.clone(), columns) {
+            let remainder = modulo_columns(numerator, denominator, &alloc);
+            let expected_remainder = if di > ni { denominator } else { numerator };
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:678]: Add a comment explaining this. 0%0 is weird.

File: crates/proof-of-sql/src/base/database/column.rs [hunk 4]
```diff
@@ -1201,4 +1214,46 @@ mod tests {
...
         assert_eq!(owned_varbinary, round_trip_owned);
     }
+
+    #[test]
+    fn we_can_get_min_scalar() {
+        assert_eq!(
+            TryInto::<i8>::try_into(ColumnType::TinyInt.min_scalar::<TestScalar>().unwrap())
+                .unwrap(),
+            i8::MIN
+        );
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/database/column.rs:1224]: This is more direct and compact:
```suggestion
        assert_eq!(ColumnType::TinyInt.min_scalar(), Some(TestScalar::from(i8::MIN)));
```

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 5]
```diff
@@ -132,6 +134,16 @@ pub(crate) fn multiply_columns<'a, S: Scalar>(
     })
 }
 
+/// Convert column to scalar slice.
+#[cfg_attr(not(test), expect(dead_code))]
+#[allow(clippy::missing_panics_doc)]
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:139]: This function could panic. Add doc explaining when that can occur.

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 6]
```diff
@@ -132,6 +134,16 @@ pub(crate) fn multiply_columns<'a, S: Scalar>(
...
+pub(crate) fn columns_to_scalar_slice<'a, S: Scalar>(
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:140]: I generally don't like this type of function that deletes the type information of a column.
Since I'm guessing you are using it in a follow-up PR, just include it with that one, so that I can see if we really need it.

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 7]
```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
...
     }
 }
+
+/// Divides two columns of data, where the data types are some signed int type(s).
+/// Note that `i128::MIN / -1`, for example, results in a value that is not contained by i128.
+/// Therefore, this value wraps around to `i128::MIN`. Division by 0 returns 0.
+/// The first slice in the tuple represents this wrapped value, whereas the second is the
+/// proper value of the quotient. Note that it is a scalar because it can represent
+///  a value (`-i128::MIN`, for example) that is out of range of the integer type.
+#[allow(clippy::missing_panics_doc)]
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:207]: Please include panics doc

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 8]
```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
...
+    }
+    (division_wrapped, division)
+}
+
+/// Modulo two columns of data, where the data types are some unsigned int type(s).
+/// Note that `i128::MIN % -1`, for example, is unusual in that `i128::MIN / -1`
+/// ordinarily returns a value that is not containe dby i128. Division wraps this operation,
+/// but modulo still returns 0 here.
+/// Division by 0 returns the numerator for modulo.
+#[allow(clippy::missing_panics_doc)]
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:246]: Please include panics doc

File: crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs [hunk 9]
```diff
@@ -185,3 +197,486 @@ pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
...
+    lhs: &Column<'a, S>,
+    rhs: &Column<'a, S>,
+    alloc: &'a Bump,
+) -> (Column<'a, S>, &'a [S]) {
+    let lhs_len = lhs.len();
+    let rhs_len = rhs.len();
+    assert!(
+        lhs_len == rhs_len,
+        "lhs and rhs should have the same length"
+    );
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_exprs/numerical_util.rs:293]: You should still test this.
https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]
